File: ast/expression/binary.go
package expression

import (
	"github.com/panda-io/micro-panda/ast/ast"
	"github.com/panda-io/micro-panda/ast/ast_types"
	"github.com/panda-io/micro-panda/token"
)

type Binary struct {
	ExpressionBase
	Left     ast.Expression
	Operator token.Token
	Right    ast.Expression
}

func (b *Binary) Validate(c ast.Context, expected ast.Type) {
	b.Left.Validate(c, expected)
	b.Right.Validate(c, b.Left.Type())

	switch b.Operator {
	case token.LeftShift, token.RightShift, token.BitXor, token.BitOr, token.BitAnd:
		b.Const = b.Left.IsConstant() && b.Right.IsConstant()
		if ast_types.IsInteger(b.Left.Type()) && ast_types.IsInteger(b.Right.Type()) {
			b.Typ = b.Left.Type()
		} else {
			c.Error(b.Left.GetPosition(), "expect integer for bit operation")
		}

	case token.Assign:
		b.Const = false
		if !b.Left.Type().Equal(b.Right.Type()) {
			c.Error(b.Left.GetPosition(), "mismatch type for binary expression")
		}
		if b.Left.IsConstant() {
			c.Error(b.Left.GetPosition(), "expect variable")
		}
		if ast_types.IsArray(b.Left.Type()) && !ast_types.IsPointer(b.Left.Type()) {
			c.Error(b.Left.GetPosition(), "array type is not assignable")
		}
		if ast_types.IsStruct(b.Left.Type()) {
			//TO-DO copy struct?
			c.Error(b.Left.GetPosition(), "struct type is not assignable")
		}

	case token.MulAssign, token.DivAssign, token.RemAssign, token.PlusAssign, token.MinusAssign:
		b.Const = false
		if !b.Left.Type().Equal(b.Right.Type()) {
			c.Error(b.Left.GetPosition(), "mismatch type for binary expression")
		}
		if b.Left.IsConstant() {
			c.Error(b.Left.GetPosition(), "expect variable")
		}
		if !(ast_types.IsNumber(b.Left.Type()) && ast_types.IsNumber(b.Right.Type())) {
			c.Error(b.Left.GetPosition(), "expect number for binary expression")
		}

	case token.LeftShiftAssign, token.RightShiftAssign, token.AndAssign, token.OrAssign, token.XorAssign:
		b.Const = false
		if !(ast_types.IsInteger(b.Left.Type()) && ast_types.IsInteger(b.Right.Type())) {
			c.Error(b.Left.GetPosition(), "expect integer for bit operation assign")
		}
		if b.Left.IsConstant() {
			c.Error(b.Left.GetPosition(), "expect variable")
		}

	case token.Or, token.And:
		b.Const = b.Left.IsConstant() && b.Right.IsConstant()
		if ast_types.IsBool(b.Left.Type()) && ast_types.IsBool(b.Right.Type()) {
			b.Typ = b.Left.Type()
		} else {
			c.Error(b.Left.GetPosition(), "mismatch type for binary expression")
		}

	case token.Less, token.LessEqual, token.Greater, token.GreaterEqual, token.Equal, token.NotEqual:
		b.Const = b.Left.IsConstant() && b.Right.IsConstant()
		if ast_types.IsNumber(b.Left.Type()) && ast_types.IsNumber(b.Right.Type()) {
			b.Typ = ast_types.TypeBool
		} else if ast_types.IsPointer(b.Left.Type()) && ast_types.IsPointer(b.Right.Type()) {
			b.Typ = ast_types.TypeBool
		} else {
			c.Error(b.Left.GetPosition(), "expect number for compare")
		}

	case token.Plus, token.Minus, token.Mul, token.Div, token.Rem:
		b.Const = b.Left.IsConstant() && b.Right.IsConstant()
		if ast_types.IsNumber(b.Left.Type()) && ast_types.IsNumber(b.Right.Type()) && b.Left.Type().Equal(b.Right.Type()) {
			b.Typ = b.Left.Type()
		} else {
			c.Error(b.Left.GetPosition(), "mismatch type for binary expression")
		}
	}
}

File: app/src/main/java/com/github/panda_io/micro_panda/ast/expression/Binary.java
package com.github.panda_io.micro_panda.ast.expression;

import com.github.panda_io.micro_panda.ast.type.Type;
import com.github.panda_io.micro_panda.ast.Context;
import com.github.panda_io.micro_panda.scanner.Token;

public class Binary extends Expression {
	public Expression left;
	public Expression right;
	public Token operator;

	public void validate(Context context, Type expected) {
		this.left.validate(context, expected);
		if (this.left.type == null) {
			return;
		}
		this.right.validate(context, this.left.type);
		if (this.right.type == null) {
			return;
		}
		if (!this.left.type.equal(this.right.type)) {
			context.addError(this.left.getOffset(), "mismatch type for binary expression");
		}

		switch (this.operator) {
			case LeftShift:
			case RightShift:
			case BitXor:
			case BitOr:
			case BitAnd:
				this.constant = this.left.isConstant() && this.right.isConstant();
				if (!this.left.type.isInteger()) {
					context.addError(this.left.getOffset(), "expect integer for bit operation");
				} else if (!this.right.type.isInteger()) {
					context.addError(this.right.getOffset(), "expect integer for bit operation");
				} else {
					this.type = this.left.type;
				}
				break;

			case Assign:
				this.constant = false;
				if (this.left.isConstant()) {
					context.addError(this.left.getOffset(), "expect variable");
				}
				if (this.left.type.isArrayWithSize()) {
					context.addError(this.left.getOffset(), "array type is not assignable");
				}
				break;

			case MulAssign:
			case DivAssign:
			case RemAssign:
			case PlusAssign:
			case MinusAssign:
				this.constant = false;
				if (this.left.isConstant()) {
					context.addError(this.left.getOffset(), "expect variable");
				}
				if (!this.left.type.isNumber()) {
					context.addError(this.left.getOffset(), "expect number for binary expression");
				}
				break;

			case LeftShiftAssign:
			case RightShiftAssign:
			case AndAssign:
			case OrAssign:
			case XorAssign:
				this.constant = false;
				if (this.left.isConstant()) {
					context.addError(this.left.getOffset(), "expect variable");
				}
				if (!this.left.type.isInteger()) {
					context.addError(this.left.getOffset(), "expect integer for binary expression");
				}
				break;

			case Or:
			case And:
				this.constant = this.left.isConstant() && this.right.isConstant();
				if (!this.left.type.isBool()) {
					context.addError(this.left.getOffset(), "expect bool binary expression");
				}
				break;

			case Less:
			case LessEqual:
			case Greater:
			case GreaterEqual:
				this.constant = this.left.isConstant() && this.right.isConstant();
				if (this.left.type.isNumber()) {
					this.type = Type.bool;
				} else {
					context.addError(this.left.getOffset(), "expect number for comparison");
				}
				break;

			case Equal:
			case NotEqual:
				this.constant = this.left.isConstant() && this.right.isConstant();
				if (this.left.type.isNumber() || this.left.type.isPointer()) {
					this.type = Type.bool;
				} else {
					context.addError(this.left.getOffset(), "expect number or pointer for comparison");
				}
				break;

			case Plus:
			case Minus:
			case Mul:
			case Div:
			case Rem:
				this.constant = this.left.isConstant() && this.right.isConstant();
				if (this.left.type.isNumber()) {
					this.type = this.left.type;
				} else {
					context.addError(this.left.getOffset(), "expect number for binary expression");
				}
				break;

			default:
				context.addError(this.left.getOffset(), "invalid operator for binary expression");
		}
	}
}
